### PR TITLE
Fix StackOverflow in Inner Join Function

### PR DESCRIPTION
**Pull request by TharushiJay**
## Purpose
This PR fixes the StackOverflow error in the inner join function of query expressions.

Fixes https://github.com/ballerina-platform/ballerina-lang/issues/39405

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/query/JoinClauseTest.java [hunk 1]
```diff
@@ -120,6 +120,12 @@ public void testOuterJoinWithNullResults() {
         BRunUtil.invoke(result, "testOuterJoin");
     }
 
+    @Test(description = "Test join clause with a large list")
+    public void testJoinClauseWithLargeList() {
+        Object values = BRunUtil.invoke(result, "testJoinClauseWithLargeList");
+        Assert.assertTrue((Boolean) values);
```
Comment (suleka96) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/query/JoinClauseTest.java:126]: Shall we assert this in `join-clause.bal` itself.

Comment (TharushiJay) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/query/JoinClauseTest.java]: Done.

File: langlib/lang.query/src/main/ballerina/types.bal [hunk 2]
```diff
@@ -340,19 +340,18 @@ class _InnerJoinFunction {
     function (_Frame _frame) returns any|error lhsKeyFunction;
     function (_Frame _frame) returns any|error rhsKeyFunction;
     _FrameMultiMap rhsFramesMap = new;
-    _Frame[]? rhsCandidates;
-    _Frame|error? lhsFrame;
     error? failureAtJoin = ();
+    stream<_Frame>? joinedFrameStream;
```
Comment (KavinduZoysa) [inline langlib/lang.query/src/main/ballerina/types.bal:344]: `joinedFramesStream`

**Review by KavinduZoysa — APPROVED** [review]

**Review by gimantha — APPROVED** [review]
LGTM

**Pull request merged**